dali/python/nvidia/dali/_utils/autoserialize.py

```diff
@@ -0,0 +1,70 @@
+# Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+
+
+def _is_marked_autoserializable(object):
+    return getattr(object, '_is_autoserialize', False)
+
+
+def _discover_autoserialize(module, visited):
+    """
+    Traverses a module tree given by the head module and
+    returns all functions that are marked with ``@autoserialize`` decorator.
+
+    :param module: Module currently searched.
+    :param visited: Paths to the ``__init__.py`` of the modules already searched.
+    :return: All functions that are marked with ``@autoserialize`` decorator.
+    """
+    assert module is not None
+    ret = []
+    module_members = inspect.getmembers(module)
+    modules = []
+    for name, path in module_members:
+        obj = getattr(module, name, None)
+        if inspect.ismodule(obj) and path not in visited:
+            modules.append(name)
+            visited.append(path)
+        elif inspect.isfunction(obj) and _is_marked_autoserializable(obj):
+            ret.append(obj)
+    for mod in modules:
+        ret.extend(_discover_autoserialize(getattr(module, mod, None), visited=visited))
+    return ret
+
+
+def invoke_autoserialize(head_module, filename):
```

Comment:
Just a nitpick, but maybe this whole file should also be present in triton module? It depends on the decorator that's already there.

Reply:
I could also see, that the `_is_autoserialize` is not the only attribute we might want to search for. In this case, adding a new one will be a matter of changing some names. In this case I believe it fits better here. Also, I believe it's better to have one `_utils` module, then several of them spread over. For these reasons, I'd go for keeping it here, but if you insist, I'll move it.

Reply:
I'm ok, that was just a question.